### PR TITLE
Add document of function torch.as_strided

### PR DESCRIPTION
**Pull request by kianasun**
Documentation of `torch.as_strided` and `Tensor.as_strided` is missing. As mentioned in #9886

### CODE REVIEW

**Review by soumith — APPROVED** [review]

File: torch/_torch_docs.py [hunk 1]
```diff
@@ -469,6 +469,35 @@ def merge_dicts(*dicts):
     True
 """)
 
+add_docstr(torch.as_strided,
+           r"""
+as_strided(input, size, stride, storage_offset=0) -> Tensor
+
+Create a `torch.Tensor` from an existing `torch.Tensor` :attr:`input` with
```
Comment (SsnL) [inline torch/_torch_docs.py]: I think it is important to mention that this creates a view of an existing tensor. :)

Comment (kianasun) [reply torch/_torch_docs.py]: You are right. Will ``Create a view of an existing torch.Tensor ...`` be better?

Comment (SsnL) [reply torch/_torch_docs.py]: Yes, that sounds great! Thank you!

Comment (SsnL) [reply torch/_torch_docs.py]: Also, could you also add an entry to the method version of it at https://github.com/pytorch/pytorch/blob/f2f80744be347fab6921fc55181b55dc86c7e07b/torch/_tensor_docs.py? It should be easy and could just refer to `torch.as_strided`.

Comment (SsnL) [reply torch/_torch_docs.py]: Actually I think `torch.as_strided` probably should not exist. Only the method version should, similar to `resize`. 

cc @gchanan who probably should have better opinion on this.

Comment (gchanan) [reply torch/_torch_docs.py]: I think it's fine, `torch.as_tensor` exists (because it needs to), so other `as_` functions seem like fair game.

Comment (kianasun) [reply torch/_torch_docs.py]: Cool. I will update both soon.

**Review by ssnl — APPROVED** [review]
Thank you!

**Review by zou3519 — CHANGES_REQUESTED** [review]
Wait, I thought it was intentional that torch.as_strided doesn't have documentation because it is dangerous and that users shouldn't be using it. Can we at least add a big red warning that this is dangerous?

**Comment by kianasun** [comment]
> Wait, I thought it was intentional that torch.as_strided doesn't have documentation because it is dangerous and that users shouldn't be using it. Can we at least add a big red warning that this is dangerous?

I think we can do that. Do we need to write some specific reasons why this is dangerous?

**Comment by zou3519** [comment]
> > Wait, I thought it was intentional that torch.as_strided doesn't have documentation because it is dangerous and that users shouldn't be using it. Can we at least add a big red warning that this is dangerous?
> 
> I think we can do that. Do we need to write some specific reasons why this is dangerous?

Things we should write:
- Can lead to the access of memory out of bounds, resulting in garbage data
- You should only use this function if you know what you're doing. A lot of pytorch functions are implemented internally with this.
- Should also have the same warning as in `tensor.expand`: https://pytorch.org/docs/master/tensors.html#torch.Tensor.expand

**Comment by SsnL** [comment]
I think that is is not that dangerous. You can’t access data outside the underlying storage. The important thing is to mention that this returns a view. We can check what np says about their counterpart in the docs.

**Comment by zou3519** [comment]
> I think that is is not that dangerous. You can’t access data outside the underlying storage. The important thing is to mention that this returns a view. We can check what np says about their counterpart in the docs.

https://docs.scipy.org/doc/numpy-1.13.0/reference/generated/numpy.lib.stride_tricks.as_strided.html for reference @kianasun

**Comment by kianasun** [comment]
@zou3519 Updated. :)

File: torch/_torch_docs.py [hunk 2]
```diff
@@ -469,6 +469,48 @@ def merge_dicts(*dicts):
...
+add_docstr(torch.as_strided,
+           r"""
+as_strided(input, size, stride, storage_offset=0) -> Tensor
+
+Create a view of an existing `torch.Tensor` :attr:`input` with specified
+:attr:`size`, :attr:`stride` and :attr:`storage_offset`.
+
+.. warning::
+    :func:`as_strided` manipulates the internal data structure of tensors.
+    If done incorrectly, it can lead to the access of invalid memory and
```
Comment (zou3519) [inline torch/_torch_docs.py]: @SsnL says this isn't possible anymore. I played around with as_strided as well and noticed there are indeed checks, so maybe it isn't applicable anymore

File: torch/_torch_docs.py [hunk 3]
```diff
@@ -469,6 +469,48 @@ def merge_dicts(*dicts):
...
+    result in garbage data.
+
+    More than one element of a created tensor may refer to a single memory
+    location. As a result, in-place operations (especially ones that are
+    vectorized) may result in incorrect behavior. If you need to write to
+    the tensors, please clone them first.
+
+    You should only use this function if you know what you're doing. A lot
```
Comment (zou3519) [inline torch/_torch_docs.py]: > You should only use this function if you know what you're doing.

I know I suggested this, but this sounds bad. Something like the following would be better:

"Many PyTorch functions that return a view on a tensor are internally implemented with this function. Those functions, like torch.expand, are easier to read and are therefore more advisable to use"

File: torch/_torch_docs.py (hunk 2 above)
Comment (kianasun) [reply torch/_torch_docs.py]: Ok. I will remove the first paragraph.

**Review by xuhdev — COMMENTED** [review]
There is also an inplace version `as_strided_`, you probably wanna document it as well (in `docs/source/tensors.rst`). Document test skipping should also be deleted:

https://github.com/pytorch/pytorch/blob/7d055c21b31bfd8a3383b13af938d39239327d03/test/test_torch.py#L196-L197

**Comment by SsnL** [comment]
I wouldn't document the inplace version.. `t_` gives us enough trouble.

**Comment by xuhdev** [comment]
@SSnL Why don't we unexpose that function if we do not want user to use it (this should probably discussed in a different thread)? Leaving a function undocumented doesn't prevent adventurous users from using it.

**Comment by SsnL** [comment]
@xuhdev We should hide it IMHO, just like many other things. It's just that no one has done it. It would probably require a deprecation cycle and fix the internal callsites.

**Review by zou3519 — APPROVED** [review]

**Pull request closed without merging**